fvm/evm/emulator/state/delta.go

```diff
@@ -0,0 +1,414 @@
+func NewDeltaView(parent types.ReadOnlyView) *DeltaView {
+	return &DeltaView{
+		parent: parent,
+
+		dirtyAddresses:      make(map[gethCommon.Address]interface{}, 0),
+		created:             make(map[gethCommon.Address]interface{}, 0),
+		suicided:            make(map[gethCommon.Address]interface{}, 0),
+		balances:            make(map[gethCommon.Address]*big.Int, 0),
+		nonces:              make(map[gethCommon.Address]uint64, 0),
+		codes:               make(map[gethCommon.Address][]byte, 0),
+		codeHashes:          make(map[gethCommon.Address]gethCommon.Hash, 0),
+		dirtySlots:          make(map[types.SlotAddress]interface{}, 0),
+		states:              make(map[types.SlotAddress]gethCommon.Hash, 0),
+		transient:           make(map[types.SlotAddress]gethCommon.Hash, 0),
+		accessListAddresses: make(map[gethCommon.Address]interface{}, 0),
+		accessListSlots:     make(map[types.SlotAddress]interface{}, 0),
+		logs:                make([]*gethTypes.Log, 0),
+		preimages:           make(map[gethCommon.Hash][]byte, 0),
+
+		// for refund we just copy the data
+		refund: parent.GetRefund(),
+	}
+}
+
+// NewChildView constructs a new delta view having the current view as parent
+func (d *DeltaView) NewChildView() *DeltaView {
+	return NewDeltaView(d)
+}
+
+// Exist returns true if address exists
+//
+// it also returns true for both newly created accounts or accounts that has been flagged for deletion
+func (d *DeltaView) Exist(addr gethCommon.Address) (bool, error) {
+	_, found := d.created[addr]
+	if found {
+		return true, nil
+	}
+	_, found = d.suicided[addr]
+	if found {
+		return true, nil
+	}
+	return d.parent.Exist(addr)
+}
+
+// CreateAccount creates a new account for the given address
+func (d *DeltaView) CreateAccount(addr gethCommon.Address) error {
+	d.created[addr] = struct{}{}
+	// flag the address as dirty
+	d.dirtyAddresses[addr] = struct{}{}
+	return nil
+}
+
+// IsCreated returns true if address has been created in this tx
+func (d *DeltaView) IsCreated(addr gethCommon.Address) bool {
+	_, found := d.created[addr]
+	if found {
+		return true
+	}
+	return d.parent.IsCreated(addr)
+}
+
+// HasSuicided returns true if address has been flagged for deletion
+func (d *DeltaView) HasSuicided(addr gethCommon.Address) bool {
+	_, found := d.suicided[addr]
+	if found {
+		return true
+	}
+	return d.parent.HasSuicided(addr)
+}
+
+// Suicide sets a flag to delete the account at the end of transaction
+func (d *DeltaView) Suicide(addr gethCommon.Address) (bool, error) {
+	// if it doesn't exist, return false
+	exists, err := d.Exist(addr)
+	if err != nil {
+		return false, err
+	}
+	if !exists {
+		return false, nil
+	}
+	// flag the account for deletion
+	d.suicided[addr] = struct{}{}
+
+	// flag the address as dirty
+	d.dirtyAddresses[addr] = struct{}{}
+	return true, nil
+}
+
+// GetBalance returns the balance of the given address
+func (d *DeltaView) GetBalance(addr gethCommon.Address) (*big.Int, error) {
+	if d.HasSuicided(addr) {
```

Comment:
could the case where the acc is suicided and then recreated exist?

Reply:
hmm, I don't know if that's possible during a single transaction, I need to investigate more.

Reply:
I checked again the logic on Geth StateDB, I updated the code to reflect that behaviour and added a test at [7a570cb](https://github.com/onflow/flow-go/pull/5167/commits/7a570cb619a2ab8d7af28c011038c4a2fcd76f88) . I also think they are going to depreciate account destruction at some point in the future.